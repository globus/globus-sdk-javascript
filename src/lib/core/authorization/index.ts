/**
 * @module Authorization
 * @description Provides modules for interacting with Globus-related authorization contexts in your application.
 * @example
 * import { authorization } from "globus/sdk";
 * const manager = authorization.create(...);
 */
import {
  AuthorizationManager,
  type AuthorizationManagerConfiguration,
} from './AuthorizationManager.js';

import * as tokens from './tokens.js';

export * as tokens from './tokens.js';

/**
 * @deprecated This will be removed in a future release. Use an instance of the AuthorizationManager instead.
 */
export const { getTokenForScope } = tokens;

/**
 * Create an instance of the {@link AuthorizationManager}.
 */
export function create(configuration: AuthorizationManagerConfiguration) {
  return new AuthorizationManager(configuration);
}
