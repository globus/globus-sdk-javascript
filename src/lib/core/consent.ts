import { getStorage } from "./storage/index.js";
import type { Token, TokenResponse } from "../services/auth/index.js";

import ITokenResponse from "js-pkce/dist/ITokenResponse";

function isValidToken(check: unknown): check is Token {
  const maybe = check as Token;
  return Boolean(maybe.token_type && maybe.access_token);
}

function store(token: ITokenResponse) {
  token.scope.split(" ").forEach((scope) => {
    getStorage().set(scope, token);
  });
}

export function addTokenResponse(token: ITokenResponse | TokenResponse) {
  store(token);
  if ("other_tokens" in token) {
    token.other_tokens.forEach(store);
  }
}

export function getTokenForScope(scope: string) {
  const token = getStorage().get(scope);
  if (!token || !isValidToken(token)) {
    return null;
  }
  return `${token.token_type} ${token.access_token}`;
}

export function reset() {
  getStorage().clear();
}
