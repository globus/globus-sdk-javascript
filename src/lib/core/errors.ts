export class EnvironmentConfigurationError extends Error {
  override name = 'EnvironmentConfigurationError';

  constructor(variable: string, value: unknown) {
    super();
    // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
    this.message = `Invalid configuration value provided for ${variable} (${value}).`;
  }
}

export type WellFormedError = {
  code: string;
  message: string;
};

export function isErrorWellFormed(test: unknown): test is WellFormedError {
  return typeof test === 'object' && test !== null && 'code' in test && 'message' in test;
}

export type ConsentRequiredError = {
  code: 'ConsentRequired';
  required_scopes: string[];
  [key: string]: unknown;
};

export function isConsentRequiredError(test: unknown): test is ConsentRequiredError {
  return (
    isErrorWellFormed(test) &&
    test.code === 'ConsentRequired' &&
    'required_scopes' in test &&
    Array.isArray(test.required_scopes)
  );
}

export type AuthorizationRequirementsError = {
  authorization_parameters: {
    session_message: string;
    session_required_identities: string[];
    session_required_mfa: boolean;
    session_required_single_domain: string[];
  };
  [key: string]: unknown;
};

export function isAuthorizationRequirementsError(
  test: unknown,
): test is AuthorizationRequirementsError {
  return (
    isErrorWellFormed(test) &&
    'authorization_parameters' in test &&
    typeof test.authorization_parameters === 'object' &&
    test.authorization_parameters !== null
  );
}
