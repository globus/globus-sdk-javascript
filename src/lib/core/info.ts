/**
 * This module is mostly intended for internal use, but can be helpful
 * identifying information about the SDK package you are using at runtime.
 */
import pkg from '../../../package.json';

export type Version = string;

/**
 * The version of the @globus/sdk package that is in use.
 */
export const VERSION: Version = pkg.version;

export type Info = {
  product: string;
  version: Version;
};

/**
 * The client information identifier for this package.
 */
export const CLIENT_INFO: Info = {
  product: 'javascript-sdk',
  version: VERSION,
};

const INFOS_SEPERATOR = ';';
const INFO_ITEM_SEPARATOR = ',';

let INFOS: Info[] = [CLIENT_INFO];

/**
 * Exported for test purposes only.
 * @private
 */
export function toString(info: Info | Info[]) {
  const infos = Array.isArray(info) ? info : [info];
  return infos
    .map((i) =>
      Object.entries(i)
        .map(([key, value]) => `${key}=${value}`)
        .join(INFO_ITEM_SEPARATOR),
    )
    .join(INFOS_SEPERATOR);
}
/**
 * Add a client information identifier to the existing SDK information.
 */
export function addClientInfo(info: Info) {
  INFOS = INFOS.concat(info);
}
/**
 * Get the current client information as a string.
 */
export function getClientInfo(): string {
  return toString(INFOS);
}

let ENABLED = true;
/**
 * Disable the client information header from being included in requests (enabled by default).
 */
export function disable() {
  ENABLED = false;
}

/**
 * Enables the client information header to be included in requests.
 */
export function enable() {
  ENABLED = true;
}

/**
 * Whether or not the client information header should be sent with requests.
 */
export function isEnabled() {
  return ENABLED;
}

/**
 * @private
 */
export const CLIENT_INFO_HEADER = `X-Globus-ClientInfo`;
