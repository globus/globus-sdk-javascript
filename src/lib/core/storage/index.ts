import { MemoryStorage } from './memory.js';
import { LocalStorage } from './local-storage.js';
/**
 * Acts as a basic wrapper for storage layers to make their surface APIs consistent.
 */
export interface StorageSystem {
  get(key: string): unknown | undefined;
  set(key: string, value: unknown): void;
  remove(key: string): void;
  clear(): void;
}

let storage: undefined | StorageSystem;
type StorageOptions =
  | 'localStorage'
  | 'memory'
  | {
      new (): StorageSystem;
    };

/**
 * Returns the active storage system or creates an instance for the running process.
 */
export function createStorage(storageType: StorageOptions = 'memory'): StorageSystem {
  if (!storage) {
    let Factory;
    if (storageType === 'localStorage') {
      Factory = LocalStorage;
    } else if (storageType === 'memory') {
      Factory = MemoryStorage;
    } else {
      Factory = storageType;
    }
    storage = new Factory();
  }
  return storage;
}

export default createStorage;

export function getStorage() {
  if (!storage) {
    throw Error('You must create a storage system.');
  }
  return storage;
}

/**
 * A private method for resetting the storage system. This is primarily used to reset
 * the storage system during testing.
 * @private
 */
// eslint-disable-next-line no-underscore-dangle, @typescript-eslint/naming-convention
export function __reset() {
  storage = undefined;
}
