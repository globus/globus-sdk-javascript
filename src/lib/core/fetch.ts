/* eslint-disable no-underscore-dangle */
import { fetch } from './internals/fetch.js';
import { getTokenForScope } from './consent.js';

type GlobusScope = string;

/**
 * Properties that will be passed to underlying `fetch` calls to override behavior.
 * @see https://developer.mozilla.org/en-US/docs/Web/API/fetch#parameters
 */
export type FetchOverrides =
  | (Omit<RequestInit, 'headers'> & {
      headers?: Record<string, string>;
      /**
       * Provide an implementation of `fetch` to be used.
       * This is currently **not** advertised since we do not dynamically import `cross-fetch` yet.
       * @private
       */
      __callable?: typeof fetch;
    })
  | undefined;

export function fetchWithScope(
  scope: GlobusScope,
  input: RequestInfo | URL,
  fetchOverrides: FetchOverrides = {},
) {
  const headers = fetchOverrides.headers || {};
  /**
   * If an `Authorization` override header was provided, we skip any
   * sort of lookup and use the provided value.
   */
  if (!headers?.['Authorization']) {
    const token = getTokenForScope(scope);
    if (token) {
      headers['Authorization'] = token;
    }
  }
  /**
   * If `Content-Type` header was not provided, and there is a body, we assume it is JSON.
   */
  if (!headers?.['Content-Type'] && fetchOverrides?.body) {
    headers['Content-Type'] = 'application/json';
  }

  /**
   * If a `__callable` override was provided, we use it instead of the default `fetch`.
   */
  if (fetchOverrides?.__callable) {
    return fetchOverrides.__callable(input, {
      ...fetchOverrides,
      headers,
    });
  }

  return fetch(input, {
    ...fetchOverrides,
    headers,
  });
}
