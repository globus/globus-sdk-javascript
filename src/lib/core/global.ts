import * as AUTH from '../services/auth/config.js';
import * as TRANSFER from '../services/transfer/config.js';
import * as FLOWS from '../services/flows/config.js';
import * as GROUPS from '../services/groups/config.js';
import * as SEARCH from '../services/search/config.js';
import * as TIMER from '../services/timer/config.js';
import * as COMPUTE from '../services/compute/config.js';

import { EnvironmentConfigurationError } from './errors.js';
import { SDKOptions } from '../services/types.js';

function getRuntime() {
  return typeof window !== 'undefined' ? window : process;
}

function isBrowser(runtime: Window | NodeJS.Process): runtime is Window {
  return typeof window === typeof runtime;
}

function env<T>(key: string, fallback: T): T {
  const runtime = getRuntime();
  let envConfiguration;
  if (isBrowser(runtime)) {
    envConfiguration = runtime;
  } else {
    envConfiguration = runtime.env;
  }
  if (key in envConfiguration) {
    return (envConfiguration as Record<typeof key, T>)[key];
  }
  return fallback;
}

/**
 * Handlers for: GLOBUS_SDK_ENVIRONMENT
 */
export const ENVIRONMENTS = {
  PRODUCTION: 'production',
  PREVIEW: 'preview',
  STAGING: 'staging',
  SANDBOX: 'sandbox',
  INTEGRATION: 'integration',
  TEST: 'test',
} as const;

export type Environment = (typeof ENVIRONMENTS)[keyof typeof ENVIRONMENTS];

export function getEnvironment(): Environment {
  const environment = env<Environment>('GLOBUS_SDK_ENVIRONMENT', ENVIRONMENTS.PRODUCTION);
  if (!environment || !Object.values(ENVIRONMENTS).includes(environment)) {
    throw new EnvironmentConfigurationError('GLOBUS_SDK_ENVIRONMENT', environment);
  }
  return environment;
}

export const SERVICES = {
  [AUTH.ID]: AUTH.ID,
  [TRANSFER.ID]: TRANSFER.ID,
  [FLOWS.ID]: FLOWS.ID,
  [GROUPS.ID]: GROUPS.ID,
  [SEARCH.ID]: SEARCH.ID,
  [TIMER.ID]: TIMER.ID,
  [COMPUTE.ID]: COMPUTE.ID,
};

export type Service = keyof typeof SERVICES;

export const SERVICE_HOSTS: Record<Service, Partial<Record<Environment, string>>> = {
  [AUTH.ID]: AUTH.HOSTS,
  [TRANSFER.ID]: TRANSFER.HOSTS,
  [FLOWS.ID]: FLOWS.HOSTS,
  [GROUPS.ID]: GROUPS.HOSTS,
  [SEARCH.ID]: SEARCH.HOSTS,
  [TIMER.ID]: TIMER.HOSTS,
  [COMPUTE.ID]: COMPUTE.HOSTS,
};

/**
 * Get the computed SDK options based on the runtime.
 * This should be used any time we're referencing the SDK options in
 * methods to ensure we're including any global overrides.
 */
export function getSDKOptions(options?: SDKOptions) {
  let globalOptions = env<string | SDKOptions>('GLOBUS_SDK_OPTIONS', {});
  if (typeof globalOptions === 'string') {
    globalOptions = JSON.parse(globalOptions) as SDKOptions;
  }
  return {
    ...globalOptions,
    ...options,
    fetch: {
      ...globalOptions?.fetch,
      ...options?.fetch,
      options: {
        ...globalOptions?.fetch?.options,
        ...options?.fetch?.options,
        headers: {
          ...globalOptions?.fetch?.options?.headers,
          ...options?.fetch?.options?.headers,
        },
      },
    },
  };
}

/**
 * Handlers for: GLOBUS_SDK_VERIFY_SSL
 * Since disabling SSL is at least not-recommended, we consider
 * this value to always be true, but provide a warning when it set
 * to one of the falsey values for informational purposes.
 *
 * Taking direction from `globus-sdk-python` for possible false values
 * @see https://github.com/globus/globus-sdk-python/blob/18eced9c12e2ec41745d1be183148845198b999c/src/globus_sdk/config/env_vars.py#L20
 */
export function getVerifySSL(): boolean {
  const verifySSLTemp = env<string>('GLOBUS_SDK_VERIFY_SSL', 'true').toLowerCase();
  if (['n', 'no', 'f', 'false', 'off', '0'].includes(verifySSLTemp)) {
    console.warn(
      'Setting GLOBUS_SDK_VERIFY_SSL to false is disallowed in the Globus Javascript SDK. It will always true in this context',
    );
  }
  return true;
}

/**
 * Handlers for: GLOBUS_SDK_HTTP_TIMEOUT
 */
export function getHttpTimeout() {
  const timeout = Number(env<string | number>('GLOBUS_SDK_HTTP_TIMEOUT', 60));
  if (timeout === -1) {
    return null;
  }
  return timeout;
}

export function getServiceHost(service: Service, environment: Environment = getEnvironment()) {
  return SERVICE_HOSTS[service][environment];
}

export function getServiceBaseUrl(service: Service, environment: Environment = getEnvironment()) {
  const host = getServiceHost(service, environment);
  return env(`GLOBUS_SDK_SERVICE_URL_${service}`, host ? `https://${host}` : undefined);
}
