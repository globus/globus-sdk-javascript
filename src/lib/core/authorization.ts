/**
 * @description Session and authorization management via PKCE.
 * @group Core
 * @module
 * @experimental
 */
import PKCE from "js-pkce";

import type IConfig from "js-pkce/dist/IConfig";
import type ITokenResponse from "js-pkce/dist/ITokenResponse";
import { getAuthorizationEndpoint, getTokenEndpoint } from "../services/auth";
import {
  addTokenResponse,
  getTokenForScope,
  reset as resetTokens,
} from "./consent";

import { createStorage, StorageSystem } from "./storage";

let storage: StorageSystem;

type PKCEConfiguration = {
  client_id?: IConfig["client_id"];
  requested_scopes: IConfig["requested_scopes"];
  redirect_uri: IConfig["redirect_uri"];
};

/**
 * @experimental
 */
export class PKCEAuthorization {
  #pkce: PKCE;
  #configuration: IConfig;
  #response: ITokenResponse | undefined;

  constructor(configuration: PKCEConfiguration) {
    storage = createStorage("localStorage");
    if (!configuration.client_id) {
      throw new Error("You must provide a `client_id`.");
    }
    this.#configuration = {
      client_id: configuration.client_id,
      authorization_endpoint: getAuthorizationEndpoint(),
      token_endpoint: getTokenEndpoint(),
      ...configuration,
    };
    this.#pkce = this.#createPKCEInstance();
  }

  #createPKCEInstance() {
    return new PKCE(this.#configuration);
  }

  #resetPKCE() {
    /**
     * Resets js-pkce state
     * @see https://github.com/bpedroza/js-pkce/blob/master/src/PKCE.ts
     */
    sessionStorage.removeItem("pkce_state");
    sessionStorage.removeItem("pkce_code_verifier");
  }

  reset() {
    this.#resetPKCE();
    resetTokens();
  }

  redirect() {
    this.reset();
    window.location.replace(this.#pkce.authorizeUrl());
  }

  async handleCodeRedirect(options = { removeStateAndReplaceLocation: true }) {
    const url = new URL(window.location.href);
    const params = new URLSearchParams(url.search);

    if (!params.get("code")) return;
    const response = await this.#pkce.exchangeForAccessToken(url.toString());
    this.#response = response;
    addTokenResponse(this.#response);
    // Remove PKCE-state from the URL since we have a token.
    if (options.removeStateAndReplaceLocation) {
      params.delete("code");
      params.delete("state");
      url.search = params.toString();
      window.location.replace(url);
    }
  }

  hasToken(): boolean {
    return this.#configuration.requested_scopes
      .split(" ")
      .every((scope) => Boolean(getTokenForScope(scope)));
  }

  revoke() {
    this.reset();
  }
}

/**
 * @experimental
 */
export function pkce(configuration: PKCEConfiguration) {
  return new PKCEAuthorization(configuration);
}

/**
 * @experimental
 */
export function incremental(configuration: PKCEConfiguration) {
  return new PKCEAuthorization(configuration);
}
