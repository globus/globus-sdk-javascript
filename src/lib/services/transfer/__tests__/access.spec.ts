import { access } from '../index';
import { createStorage } from '../../../core/storage';

import { mirror } from '../../../../__mocks__/handlers';

describe('task', () => {
  beforeEach(() => {
    createStorage('memory');
  });

  test('getAll', async () => {
    const {
      req: { url, method, headers },
    } = await mirror(await access.getAll('example-endpoint-id'));
    expect({
      url,
      method,
      headers,
    }).toMatchSnapshot();
  });

  test('get', async () => {
    const {
      req: { url, method, headers },
    } = await mirror(
      await access.get({ endpoint_xid: 'example-task-id', id: 'example-access-rule-id' }),
    );
    expect({
      url,
      method,
      headers,
    }).toMatchSnapshot();
  });

  test('update', async () => {
    const {
      req: { url, method, headers, json },
    } = await mirror(
      await access.update(
        { endpoint_xid: 'example-task-id', id: 'example-access-rule-id' },
        {
          payload: {
            permissions: 'rw',
          },
        },
      ),
    );
    expect({
      url,
      method,
      headers,
      json,
    }).toMatchSnapshot();
  });

  test('remove', async () => {
    const {
      req: { url, method, headers, json },
    } = await mirror(
      await access.remove({ endpoint_xid: 'example-task-id', id: 'example-access-rule-id' }),
    );
    expect({
      url,
      method,
      headers,
      json,
    }).toMatchSnapshot();
  });
});
