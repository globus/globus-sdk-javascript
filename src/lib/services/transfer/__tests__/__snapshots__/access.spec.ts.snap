// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`task get 1`] = `
{
  "headers": {
    "accept": "*/*",
    "accept-encoding": "gzip,deflate",
    "connection": "close",
    "host": "transfer.api.globusonline.org",
    "user-agent": "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)",
  },
  "method": "GET",
  "url": "https://transfer.api.globusonline.org/v0.10/endpoint/example-task-id/access/example-access-rule-id",
}
`;

exports[`task getAll 1`] = `
{
  "headers": {
    "accept": "*/*",
    "accept-encoding": "gzip,deflate",
    "connection": "close",
    "host": "transfer.api.globusonline.org",
    "user-agent": "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)",
  },
  "method": "GET",
  "url": "https://transfer.api.globusonline.org/v0.10/endpoint/example-endpoint-id/access_list",
}
`;

exports[`task remove 1`] = `
{
  "headers": {
    "accept": "*/*",
    "accept-encoding": "gzip,deflate",
    "connection": "close",
    "host": "transfer.api.globusonline.org",
    "user-agent": "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)",
  },
  "json": undefined,
  "method": "DELETE",
  "url": "https://transfer.api.globusonline.org/v0.10/endpoint/example-task-id/access/example-access-rule-id",
}
`;

exports[`task update 1`] = `
{
  "headers": {
    "accept": "*/*",
    "accept-encoding": "gzip,deflate",
    "connection": "close",
    "content-length": "20",
    "content-type": "application/json",
    "host": "transfer.api.globusonline.org",
    "user-agent": "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)",
  },
  "json": {
    "permissions": "rw",
  },
  "method": "PUT",
  "url": "https://transfer.api.globusonline.org/v0.10/endpoint/example-task-id/access/example-access-rule-id",
}
`;
